# Answering Business Questions Using SQL

In this project, we will use SQL to answer business questions. The database we will use is the Chinook database, a widely used database that represent a digital media store. You can find the database [here](https://github.com/lerocha/chinook-database?tab=readme-ov-file). 

Studying the schema diagram provided with the Chinook database is an important step before starting this project. You can find it in the ``` README ``` for the GitHub repository, a link for which is provided above.

## Setting Up SQL Environment and Linking the SQLite Database

In [1]:
%%capture
!pip install ipython-sql

In [2]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

## Data Overview

We will first study the columns in our database.

In [3]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

 * sqlite:///chinook.db
Done.


[('albums', 'table'),
 ('sqlite_sequence', 'table'),
 ('artists', 'table'),
 ('customers', 'table'),
 ('employees', 'table'),
 ('genres', 'table'),
 ('invoices', 'table'),
 ('invoice_items', 'table'),
 ('media_types', 'table'),
 ('playlists', 'table'),
 ('playlist_track', 'table'),
 ('tracks', 'table'),
 ('sqlite_stat1', 'table')]

### Selecting New Albums to Purchase

The store needs to make a data-driven decision about which albums they should purchase. They need to know which albums will sell well in the United States. We will write a query to find the number and percentage of tracks sold in USA in different genres to determine which genres are the most popular.

For this, we will first create a temporary table called ``` genre_usa_tracks ``` to extract all the tracks sold in the US with their respective genres.

In [4]:
%%sql

WITH genre_usa_tracks AS
    (
        SELECT ii.InvoiceLineId, t.TrackId, g.Name AS Genre, i.BillingCountry
        FROM invoices AS i
        JOIN invoice_items AS ii ON i.InvoiceId = ii.InvoiceId
        JOIN tracks AS t ON ii.TrackId = t.TrackId
        JOIN genres AS g ON t.GenreId = g.GenreId
        WHERE i.BillingCountry = "USA"
)

SELECT Genre, 
COUNT(*) AS num_tracks_sold,
COUNT(*)/(SELECT CAST(COUNT(*) AS float) FROM genre_usa_tracks) * 100 AS percentage_tracks --cast to float because int division results in 0 (rounded off)
FROM genre_usa_tracks
GROUP BY Genre
ORDER BY 2 DESC

 * sqlite:///chinook.db
Done.


[('Rock', 157, 31.781376518218625),
 ('Latin', 91, 18.421052631578945),
 ('Metal', 64, 12.955465587044534),
 ('Alternative & Punk', 50, 10.121457489878543),
 ('Jazz', 22, 4.4534412955465585),
 ('Blues', 15, 3.0364372469635628),
 ('TV Shows', 14, 2.834008097165992),
 ('R&B/Soul', 12, 2.42914979757085),
 ('Comedy', 8, 1.6194331983805668),
 ('Classical', 8, 1.6194331983805668),
 ('Bossa Nova', 7, 1.417004048582996),
 ('Reggae', 6, 1.214574898785425),
 ('Drama', 6, 1.214574898785425),
 ('Sci Fi & Fantasy', 5, 1.0121457489878543),
 ('Pop', 5, 1.0121457489878543),
 ('Alternative', 5, 1.0121457489878543),
 ('Soundtrack', 4, 0.8097165991902834),
 ('Hip Hop/Rap', 4, 0.8097165991902834),
 ('Heavy Metal', 4, 0.8097165991902834),
 ('Rock And Roll', 3, 0.6072874493927125),
 ('Easy Listening', 3, 0.6072874493927125),
 ('Science Fiction', 1, 0.20242914979757085)]

We can see that the top 2 genres with the most tracks sold in the US are Rock and Latin. The store should invest in albums from these genres, with a focus on Rock, since it has the highest percentage of tracks sold (nearly 32%). 

Metal and Alternative & Punk also show some potential, with track sales at approximately 12% and 10%, respectively.

## Analyzing Employee Sales Performance

The store has a sales representative assigned to each of its customers. They need to find out which of their employees has managed to make the most sales. We will calculate the total dollars each salesperson has made in sales. To get a better idea of employees' sales performances, we will calculate the average amount of sales in dollars per customer.

The ``` HireDate ``` column is relevant to us since an employee's total sales depends on when they were hired.

In [5]:
%%sql

SELECT 
    e.FirstName || ' ' || e.LastName AS Name,
    e.HireDate,
    COUNT(c.CustomerId) AS customer_count,
    SUM(i.Total) AS total_sales_dollars,
    SUM(i.Total) / COUNT(c.CustomerId) AS avg_dollars_per_customer
FROM 
    employees AS e
JOIN 
    customers AS c 
    ON e.EmployeeId = c.SupportRepId
JOIN 
    invoices AS i 
    ON i.CustomerId = c.CustomerId
GROUP BY 1,2

 * sqlite:///chinook.db
Done.


[('Jane Peacock', '2002-04-01 00:00:00', 146, 833.0400000000013, 5.705753424657543),
 ('Margaret Park', '2003-05-03 00:00:00', 140, 775.4000000000011, 5.538571428571436),
 ('Steve Johnson', '2003-10-17 00:00:00', 126, 720.160000000001, 5.7155555555555635)]

We can see that Steve Johnson outperforms the rest of the employees. According to his hire date, he is the newest employee, yet still has the highest average dollars per customer of 5.72, which narrowly wins over Jane Peacock's 5.71 figure, who is the oldest employee.

## Analysing Sales by Country

The store needs to know how much dollars in sales they have made to customers from different countries. We will find the total number of customers and the amount of sales in dollars for each country.

- To clean up our analysis, we will group countries with 1 customer into an ``` Other ``` category.
- For better presentation, we will sort the results by sales amount, and then country name. ``` Other ``` will appear at the bottom.

In [6]:
%%sql

WITH 
    customer_counts AS (
        SELECT
            Country,
            COUNT(*) AS num_customers
        FROM
            customers
        GROUP BY 1
        ),

    labelled_invoices AS(
        SELECT 
            CASE WHEN cc.num_customers = 1 THEN 'Other'
            ELSE c.Country 
            END AS Country,
            i.CustomerId,
            i.invoiceId,
            ii.UnitPrice,
            ii.Quantity
        FROM
            customer_counts AS cc
        JOIN
            customers AS c USING(Country)
        JOIN 
            invoices AS i USING(CustomerId)
        JOIN
            invoice_items AS ii USING(InvoiceId)
        )

SELECT 
    Country,
    COUNT(DISTINCT CustomerId) AS Customers,
    SUM(UnitPrice * Quantity) AS total_sales_dollars
FROM 
    labelled_invoices
GROUP BY 1
ORDER BY
    CASE WHEN Country = 'Other' THEN 1 ELSE 0 END, 3 DESC

 * sqlite:///chinook.db
Done.


[('USA', 13, 523.060000000003),
 ('Canada', 8, 303.96000000000134),
 ('France', 5, 195.1000000000003),
 ('Brazil', 5, 190.10000000000028),
 ('Germany', 4, 156.48),
 ('United Kingdom', 3, 112.85999999999979),
 ('Czech Republic', 2, 90.23999999999997),
 ('Portugal', 2, 77.23999999999997),
 ('India', 2, 75.25999999999999),
 ('Other', 15, 604.3000000000038)]

## Album vs. Individual Tracks (Incomplete)

In [13]:
%%sql

WITH invoice_first_tracks AS (
    SELECT 
        InvoiceId,
        MIN(TrackId) AS first_track_id
    FROM 
        invoice_items
    GROUP BY 
        1
)

SELECT
    ift.InvoiceId,
    t.TrackId,
    t.AlbumId
FROM
    invoice_first_tracks AS ift
JOIN
    tracks AS t
ON ift.first_track_id = t.TrackId


SELECT 
    t.TrackID,
    t.AlbumId,
    ii.InvoiceId
FROM
    tracks as t
JOIN
    invoice_items AS ii
USING (InvoiceId)


SELECT TrackId
WHERE ii.InvoiceId = ift.InvoiceId

 * sqlite:///chinook.db
Done.


[(1, 2),
 (2, 6),
 (3, 16),
 (4, 42),
 (5, 99),
 (6, 230),
 (7, 231),
 (8, 234),
 (9, 238),
 (10, 248),
 (11, 274),
 (12, 331),
 (13, 462),
 (14, 463),
 (15, 466),
 (16, 470),
 (17, 480),
 (18, 506),
 (19, 563),
 (20, 694),
 (21, 695),
 (22, 698),
 (23, 702),
 (24, 712),
 (25, 738),
 (26, 795),
 (27, 926),
 (28, 927),
 (29, 930),
 (30, 934),
 (31, 944),
 (32, 970),
 (33, 1027),
 (34, 1158),
 (35, 1159),
 (36, 1162),
 (37, 1166),
 (38, 1176),
 (39, 1202),
 (40, 1259),
 (41, 1390),
 (42, 1391),
 (43, 1394),
 (44, 1398),
 (45, 1408),
 (46, 1434),
 (47, 1491),
 (48, 1622),
 (49, 1623),
 (50, 1626),
 (51, 1630),
 (52, 1640),
 (53, 1666),
 (54, 1723),
 (55, 1854),
 (56, 1855),
 (57, 1858),
 (58, 1862),
 (59, 1872),
 (60, 1898),
 (61, 1955),
 (62, 2086),
 (63, 2087),
 (64, 2090),
 (65, 2094),
 (66, 2104),
 (67, 2130),
 (68, 2187),
 (69, 2318),
 (70, 2319),
 (71, 2322),
 (72, 2326),
 (73, 2336),
 (74, 2362),
 (75, 2419),
 (76, 2550),
 (77, 2551),
 (78, 2554),
 (79, 2558),
 (80, 2568),
 (81, 2594),
 (82, 2651),
 (83, 2782),
 (84, 2783),
 (85, 2786),
 (86, 2790),
 (87, 2800),
 (88, 2826),
 (89, 2883),
 (90, 3014),
 (91, 3015),
 (92, 3018),
 (93, 3022),
 (94, 3032),
 (95, 3058),
 (96, 3115),
 (97, 3246),
 (98, 3247),
 (99, 3250),
 (100, 3254),
 (101, 3264),
 (102, 3290),
 (103, 3347),
 (104, 3478),
 (105, 3479),
 (106, 3482),
 (107, 3486),
 (108, 1),
 (109, 19),
 (110, 76),
 (111, 207),
 (112, 208),
 (113, 211),
 (114, 215),
 (115, 225),
 (116, 251),
 (117, 308),
 (118, 439),
 (119, 440),
 (120, 443),
 (121, 447),
 (122, 457),
 (123, 483),
 (124, 540),
 (125, 671),
 (126, 672),
 (127, 675),
 (128, 679),
 (129, 689),
 (130, 715),
 (131, 772),
 (132, 903),
 (133, 904),
 (134, 907),
 (135, 911),
 (136, 921),
 (137, 947),
 (138, 1004),
 (139, 1135),
 (140, 1136),
 (141, 1139),
 (142, 1143),
 (143, 1153),
 (144, 1179),
 (145, 1236),
 (146, 1367),
 (147, 1368),
 (148, 1371),
 (149, 1375),
 (150, 1385),
 (151, 1411),
 (152, 1468),
 (153, 1599),
 (154, 1600),
 (155, 1603),
 (156, 1607),
 (157, 1617),
 (158, 1643),
 (159, 1700),
 (160, 1831),
 (161, 1832),
 (162, 1835),
 (163, 1839),
 (164, 1849),
 (165, 1875),
 (166, 1932),
 (167, 2063),
 (168, 2064),
 (169, 2067),
 (170, 2071),
 (171, 2081),
 (172, 2107),
 (173, 2164),
 (174, 2295),
 (175, 2296),
 (176, 2299),
 (177, 2303),
 (178, 2313),
 (179, 2339),
 (180, 2396),
 (181, 2527),
 (182, 2528),
 (183, 2531),
 (184, 2535),
 (185, 2545),
 (186, 2571),
 (187, 2628),
 (188, 2759),
 (189, 2760),
 (190, 2763),
 (191, 2767),
 (192, 2777),
 (193, 2803),
 (194, 2860),
 (195, 2991),
 (196, 2992),
 (197, 2995),
 (198, 2999),
 (199, 3009),
 (200, 3035),
 (201, 3092),
 (202, 3223),
 (203, 3224),
 (204, 3227),
 (205, 3231),
 (206, 3241),
 (207, 3267),
 (208, 3324),
 (209, 3455),
 (210, 3456),
 (211, 3459),
 (212, 3463),
 (213, 3473),
 (214, 2),
 (215, 53),
 (216, 184),
 (217, 185),
 (218, 188),
 (219, 192),
 (220, 202),
 (221, 228),
 (222, 285),
 (223, 416),
 (224, 417),
 (225, 420),
 (226, 424),
 (227, 434),
 (228, 460),
 (229, 517),
 (230, 648),
 (231, 649),
 (232, 652),
 (233, 656),
 (234, 666),
 (235, 692),
 (236, 749),
 (237, 880),
 (238, 881),
 (239, 884),
 (240, 888),
 (241, 898),
 (242, 924),
 (243, 981),
 (244, 1112),
 (245, 1113),
 (246, 1116),
 (247, 1120),
 (248, 1130),
 (249, 1156),
 (250, 1213),
 (251, 1344),
 (252, 1345),
 (253, 1348),
 (254, 1352),
 (255, 1362),
 (256, 1388),
 (257, 1445),
 (258, 1576),
 (259, 1577),
 (260, 1580),
 (261, 1584),
 (262, 1594),
 (263, 1620),
 (264, 1677),
 (265, 1808),
 (266, 1809),
 (267, 1812),
 (268, 1816),
 (269, 1826),
 (270, 1852),
 (271, 1909),
 (272, 2040),
 (273, 2041),
 (274, 2044),
 (275, 2048),
 (276, 2058),
 (277, 2084),
 (278, 2141),
 (279, 2272),
 (280, 2273),
 (281, 2276),
 (282, 2280),
 (283, 2290),
 (284, 2316),
 (285, 2373),
 (286, 2504),
 (287, 2505),
 (288, 2508),
 (289, 2512),
 (290, 2522),
 (291, 2548),
 (292, 2605),
 (293, 2736),
 (294, 2737),
 (295, 2740),
 (296, 2744),
 (297, 2754),
 (298, 2780),
 (299, 2837),
 (300, 2968),
 (301, 2969),


In [16]:
%%sql
WITH invoice_first_tracks AS (
    SELECT 
        InvoiceId,
        MIN(TrackId) AS first_track_id
    FROM 
        invoice_items
    GROUP BY 
        1
)

SELECT
    ift.InvoiceId,
    t.TrackId,
    t.AlbumId
FROM
    invoice_first_tracks AS ift
JOIN
    tracks AS t
ON ift.first_track_id = t.TrackId
LIMIT 5

 * sqlite:///chinook.db
Done.


[(1, 2, 2), (2, 6, 1), (3, 16, 4), (4, 42, 6), (5, 99, 11)]